src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenChannelAdapter.java

```diff
@@ -271,6 +288,32 @@ public int afterShutdown() {
 		return getPhase();
 	}
 
+	protected void setAttributesIfNecessary(Object record, Message<?> message) {
+		boolean needAttributes = getErrorChannel() != null
+				&& KafkaMessageDrivenChannelAdapter.this.retryTemplate == null;
+		if (needAttributes) {
+			attributesHolder.set(ErrorMessageUtils.getAttributeAccessor(null, null));
```

Comment:
? That's different logic.

`needAttributes` is true when there's an error channel and no retry template.

In that case, we need to construct the attributes.

Your logic would overwrite the attributes (retryContext) set by the retry template (via the listener).

Plus, it wouldn't set the attributes if there's no recovery callback.

I'll add javadoc to explain.